pkg/controllers/managementuser/rbac/handler_base.go

```diff
@@ -343,24 +343,31 @@ func ToLowerRoleTemplates(roleTemplates map[string]*v3.RoleTemplate) {
 	}
 }
 
-func (m *manager) gatherRoles(rt *v3.RoleTemplate, roleTemplates map[string]*v3.RoleTemplate) error {
-	err := m.gatherRolesRecurse(rt, roleTemplates)
+func (m *manager) gatherRoles(rt *v3.RoleTemplate, roleTemplates map[string]*v3.RoleTemplate, depthCounter int) error {
+	if depthCounter == 100 {
+		logrus.Warn("roletemplate has cause 100 recursions")
+	}
+	if depthCounter >= 500 {
+		return errors.New("roletemplate has recursed over 1000 times, possible circular dependency")
+	}
+	err := m.gatherRolesRecurse(rt, roleTemplates, depthCounter)
 	if err != nil {
 		return err
 	}
 	ToLowerRoleTemplates(roleTemplates)
 	return nil
 }
 
-func (m *manager) gatherRolesRecurse(rt *v3.RoleTemplate, roleTemplates map[string]*v3.RoleTemplate) error {
+func (m *manager) gatherRolesRecurse(rt *v3.RoleTemplate, roleTemplates map[string]*v3.RoleTemplate, depthCounter int) error {
 	roleTemplates[rt.Name] = rt
 
 	for _, rtName := range rt.RoleTemplateNames {
+		depthCounter++
```

Comment:
Since this depth counter is included inside of the for loop, it kind of serves as a depth and breadth counter which I'm not a super big fan of. More specifically, you could have a non-infinitely recursing role that just inherits from 1000 different roles and this code would cause an error. I would recommend incrementing this before the for loop so that this serves only as a depth counter.